### PR TITLE
CC-398 only src for phpstan, add ability to include additional paths

### PR DESCRIPTION
**Pull request by sokyrko**
Task: https://spryker.atlassian.net/browse/CC-398

### CODE REVIEW

**Review by dereuromark — CHANGES_REQUESTED** [review]
The ticket says

- "Git hooks apply only for files in a "src" folder"

The changes are invalid as is, since they still by default would check way too many folders by default.

Instead, ONLY run into src/ folder as outlined in the ticket specs.

Important:

- MUST only ever check ROOT/src/ dirs
- MUST be in sync with core tooling (which also does that this way, see travis file)

Can't we reuse the same high level commands here by the way, @stereomon  ?

**Comment by stereomon** [comment]
> "Git hooks apply only for files in a "src" folder"

This will not work for some checks we make with the provided hooks. E.g. the dependency check tool will not show mistakes if you only change a test and add/remove a dependency there. Same for changes in composer.json, codeception.yml and .travis.yml. 

AFAIK it was never a problem to also run checks on changed files in the modules root or tests directory. Anything which can be found before the CI server is running can save time. If something takes to much time to fix in a changed test the no-validation flag can be used to skip further checks.

IMO Tests should also be as clean as possible and should contain valid code. I would not disable something which (so far) didn't hurt. 

It would take a lot of time to refactor the package to only trigger the hooks for changed files in the src directory. 

> Can't we reuse the same high level commands here by the way

So far only the dependency check tool does that, all other use e.g. phpmd directly. In some cases we would get to much noise on files we did not even touched during development. So where ever possible we use the full path to a changed file instead of running checks for the whole module of a changed file.

**Comment by dereuromark** [comment]
We already discussed this: The phpstan command must be in sync with current tooling, and as such MUST only run into /src/ 
Nothing else is acceptable - otherwise please remove phpstan completely from githook tooling for now as "broken to be fixed later".

**Comment by stereomon** [comment]
Can you please describe what/why is broken with phpstan usage in tests and create a separate ticket so this one can be closed soon? 

AFAIK it works pretty well also in tests directories and is more helpful to have than not to have. High-quality test code makes changes in their less time consuming and less error-prone.

> The phpstan command must be in sync with current tooling

Back in the days, we started in a similar way with our code style check tool, it ran only locally until we were able to turn it on on the remote CI server after we made a lot fixes locally with every change we made. Maybe we will take care about the same quality in the future for our tests and if so with the hooks enabled for tests it will not hurt that much when we turn it on for the CI server someday.

**Comment by dereuromark** [comment]
It is fine of you manually run it for tests or other dirs. 
But the githook is automatically running and as such must provide the same feedback the current CI tooling and rest of automatic tooling infrastructure does.

So even if it would be nice to have tests covered, this is not acceptable as is by default for now.
The command should be fixed here to run in sync with tooling, not additional things.
That is something can be opt-in configured, though, if desired on local level.

File: src/GitHook/Command/FileCommand/PreCommit/PhpStan/PhpStanConfiguration.php [hunk 1]
```diff
@@ -35,6 +35,30 @@ public function getLevel()
...
+    {
+        return array_merge([$this->getSourceDir()], $this->getAdditionalDirs());
+    }
+
+    /**
+     * @return string
+     */
+    public function getSourceDir(): string
+    {
+        return 'src/';
```
Comment (dereuromark) [inline src/GitHook/Command/FileCommand/PreCommit/PhpStan/PhpStanConfiguration.php]: `'src' . DIRECTORY_SEPARATOR` better

**Review by dereuromark — APPROVED** [review]
For now this is fine until this can be refactored to cleanly use the config that console command is using.
Thanks for fixing this up

Refs also https://github.com/spryker/suite-nonsplit/pull/721

File: src/GitHook/Command/FileCommand/PreCommit/PhpStan/PhpStanConfiguration.php [hunk 2]
```diff
@@ -35,6 +35,30 @@ public function getLevel()
...
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getAdditionalDirs(): array
+    {
+        return $this->config['additionalDirs'] ?? [];
```
Comment (dereuromark) [inline src/GitHook/Command/FileCommand/PreCommit/PhpStan/PhpStanConfiguration.php]: You can make those all long "Directories"
also in method names.

File: src/GitHook/Command/FileCommand/PreCommit/PhpStan/PhpStanConfiguration.php [hunk 3]
```diff
@@ -40,23 +40,23 @@ public function getLevel()
...
     }
 
     /**
      * @return string
      */
-    public function getSourceDir(): string
+    public function getSourceDirectory(): string
     {
-        return 'src/';
+        return 'src';
```
Comment (dereuromark) [inline src/GitHook/Command/FileCommand/PreCommit/PhpStan/PhpStanConfiguration.php]: Why removing? We always have trailing slash, see my last message on how it is done properly.

File: src/GitHook/Helper/CommittedFilesHelper.php [hunk 4]
```diff
@@ -24,7 +24,7 @@ public function getCommittedFiles()
             $against = 'HEAD';
         }
 
-        exec('git diff-index --cached --name-status ' . $against . ' | egrep \'^(A|M)\' | awk \'{print $2;}\'', $committedFiles);
+        exec('git diff-index --name-status ' . $against . ' | egrep \'^(A|M)\' | awk \'{print $2;}\'', $committedFiles);
```
Comment (dereuromark) [inline src/GitHook/Helper/CommittedFilesHelper.php:27]: This fixes cache issues :+1:

**Review by dereuromark — APPROVED** [review]

**Pull request merged**